src/config.cpp

```diff
@@ -478,5 +479,127 @@ const std::string& CircuitConfig::getExpandedJSON() const {
+        if (element == it.end())
+            throw SonataError(fmt::format("Could not find '{}' in '{}'", name, sn));
+        buf = element->template get<Type>();
+    }
+
+    template <typename Iterator, typename Type>
+    void parseOptional(const Iterator& it, const char* name, Type& buf) const {
+        const auto element = it.find(name);
+        if (element != it.end())
+            buf = element->template get<Type>();
+    }
+
+    SimulationConfig::Run parseRun() const {
+        const auto runIt = _json.find("run");
+        if (runIt == _json.end())
+            throw SonataError("Could not find 'run' section");
+
+        SimulationConfig::Run result{};
+        parseMandatory(*runIt, "tstop", "run", result.tstop);
+        parseMandatory(*runIt, "dt", "run", result.dt);
+        return result;
+    }
+
+    SimulationConfig::Output parseOutput() const {
+        SimulationConfig::Output result{};
+        result.outputDir = "output";
+        result.spikesFile = "out.h5";
+
+        const auto outputIt = _json.find("output");
+        if (outputIt != _json.end()) {
+            parseOptional(*outputIt, "output_dir", result.outputDir);
+            parseOptional(*outputIt, "spikes_file", result.spikesFile);
+        }
+
+        result.outputDir = toAbsolute(_basePath, result.outputDir);
+
+        return result;
+    }
+
+    std::unordered_map<std::string, SimulationConfig::Report> parseReports() const {
+        std::unordered_map<std::string, SimulationConfig::Report> result;
+
+        const auto reportsIt = _json.find("reports");
+        if (reportsIt == _json.end())
+            return result;
+
+        for (auto it = reportsIt->begin(); it != reportsIt->end(); ++it) {
+            auto& report = result[it.key()];
+            auto& valueIt = it.value();
+            const auto debugStr = fmt::format("report {}", it.key());
+            parseMandatory(valueIt, "cells", debugStr, report.cells);
+            parseMandatory(valueIt, "type", debugStr, report.type);
+            parseMandatory(valueIt, "dt", debugStr, report.dt);
+            parseMandatory(valueIt, "start_time", debugStr, report.startTime);
+            parseMandatory(valueIt, "end_time", debugStr, report.endTime);
+            parseOptional(valueIt, "file_name", report.fileName);
+            if (report.fileName.empty())
+                report.fileName = it.key() + "_SONATA.h5";
+            else if (report.fileName.find(".h5") == std::string::npos)
+                report.fileName += ".h5";
+            report.fileName = toAbsolute(_basePath, report.fileName);
+        }
+
+        return result;
+    }
+
+  private:
+    const fs::path _basePath;
+    const nlohmann::json _json;
+};
+
+SimulationConfig::SimulationConfig(const std::string& content, const std::string& basePath)
+    : _jsonContent(content)
+    , _basePath(fs::absolute(basePath).lexically_normal().string()) {
+    const Parser parser(content, basePath);
```

Comment:
Not sure I see the point of this `Parser` class, but I guess you have done it this way for consistency. This seems to imply a pointless copy of `content` and `basePath`. If I was starting from a clean slate I would lean towards implementing `parse{Run,Output,Reports}()` as pure functions in an anonymous namespace. This way you could call them directly in the initializer list.

Reply:
Yes, its been done like that for consistency with the `CircuitConfig` class implementation. For `CircuitConfig`, the `Parser` class was created so that the parsing would be outside the `CircuitConfig` class itself, while the `Parser` class is also friend of `CircuitConfig` to allow access the private members. 

Its not necessary for `SimulationConfig`, although I also don't think its an expensive copy which usually would be performed just once, and also helps keep the code clean and more readable.

However, I don't have a strong preference for either implementation.